src/node/internal/internal_tls_jsstream.ts

```diff
@@ -0,0 +1,308 @@
+// in all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN
+// NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
+// DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
+// OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE
+// USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+import { ok, strictEqual, notStrictEqual } from 'node-internal:internal_assert';
+import { Socket } from 'node-internal:internal_net';
+import { ERR_STREAM_WRAP } from 'node-internal:internal_errors';
+import { Duplex } from 'node-internal:streams_duplex';
+import type {
+  SocketInfo,
+  Writer,
+  Socket as CloudflareSocket,
+} from 'node-internal:sockets';
+
+const kCurrentWriteRequest = Symbol('kCurrentWriteRequest');
+const kCurrentShutdownRequest = Symbol('kCurrentShutdownRequest');
+const kPendingShutdownRequest = Symbol('kPendingShutdownRequest');
+const kPendingClose = Symbol('kPendingClose');
+
+/* This class serves as a wrapper for when the C++ side of Node wants access
+ * to a standard JS stream. For example, TLS or HTTP do not operate on network
+ * resources conceptually, although that is the common case and what we are
+ * optimizing for; in theory, they are completely composable and can work with
+ * any stream resource they see.
+ *
+ * For the common case, i.e. a TLS socket wrapping around a net.Socket, we
+ * can skip going through the JS layer and let TLS access the raw C++ handle
+ * of a net.Socket. The flipside of this is that, to maintain composability,
+ * we need a way to create "fake" net.Socket instances that call back into a
+ * "real" JavaScript stream. JSStreamSocket is exactly this.
+ */
+export class JSStreamSocket extends Socket {
+  public stream: Duplex;
+  public [kCurrentWriteRequest]: null | unknown;
+  public [kCurrentShutdownRequest]: null | unknown;
+  public [kPendingShutdownRequest]: null | unknown;
+  public [kPendingClose]: boolean;
+
+  #openPromise: ReturnType<typeof Promise.withResolvers<SocketInfo>>;
+
+  constructor(stream: Duplex) {
+    const closePromise = Promise.withResolvers<void>();
+    const openPromise = Promise.withResolvers<SocketInfo>();
+
+    const webStream = Duplex.toWeb(stream);
+    const writable: Writer = Object.assign({}, webStream.writable, {
+      write: async (data: string | ArrayBufferView): Promise<void> => {
+        stream.write(data);
+      },
+      closed: closePromise.promise,
+      releaseLock: async (): Promise<void> => {},
+    });
+    const handle: Socket['_handle'] = {
+      reading: true,
+      bytesRead: 0,
+      bytesWritten: 0,
+      socket: {
+        startTls(): CloudflareSocket {
+          throw new Error('This should not have triggered');
```

Comment:
I know this is more of an internal error than a user facing one but the message could be improved more